File: 백준 문제 풀이 모음/class 4/1043.py
import sys
input = sys.stdin.readline

def find(parent, x) :
    if x != parent[x] :
        parent[x] = find(parent, parent[x])

    return parent[x]

def union(parent, a, b, know_truth) :
    a = find(parent, a)
    b = find(parent, b)

    if a in know_truth and b in know_truth :
        return
    
    if a in know_truth :
        parent[b] = a
    elif b in know_truth :
        parent[a] = b
    
    else :
        if a < b :
            parent[b] = a
        else :
            parent[a] = b
     

n, m = map(int, input().split()) #n=10, m=9

know_truth = list(map(int, input().split()))[1:] #[1,2,3,4]

parties = []
parent = list(range(n+1)) # parent = [0,1,2,3,4,5,6,7,8,9,10]


for _ in range(m) :
    party_info = list(map(int, input().split()))
    party_len = party_info[0] # party_len = 4
    party = party_info[1:] #party = [1,2,3,4]

    for i in range(party_len - 1) :
        union(parent, party[i], party[i+1], know_truth)

File: 백준 문제 풀이 모음/kotlin/class 4/1043.kt
import java.util.*


lateinit var parent : MutableList<Int>
lateinit var know_truth : MutableList<Boolean>

fun main() = with(System.`in`.bufferedReader()){
    val (n, m) = readLine().split(' ').toList().map { it.toInt() }
    // n = 사람 수, m = 파티 수
    parent = MutableList<Int>(n+1) { it }

    know_truth = MutableList<Boolean>(n+1) { false }

    val input = StringTokenizer(readLine())
    val num = input.nextToken().toInt()
    repeat(num) {
        know_truth[input.nextToken().toInt()] = true
    }

    val party = MutableList<MutableList<Int>>(m) { MutableList<Int> (0) { 0 } }
    repeat(m) {i ->
        val input = StringTokenizer(readLine())
        val num2 = input.nextToken().toInt() // 파티에 참여한 사람 수
        for (j in 0 until num2) {
            party[i].add(input.nextToken().toInt())

            if (j > 0) union(party[i][j], party[i][j-1])
        }
    }


    var ans = 0

    party.forEach {
        var state = true
        for (i in 0 until it.size) {
            if (!know_truth[find(it[i])])  {
                state = false
                break
            }
        }
        if (!state) {
            ans ++
        }
    }
    print(ans)
//    val result = mutableListOf<Int>()
//    for (i in 0 until know_truth.size) {
//        if (know_truth[i]) result.add(i)
//    }
//
//    var count = 0
//    repeat (m) {i ->
//        for (j in 0 until party[i].size) {
//            if (result.contains(party[i][j])) {
//                count ++
//                break
//            }
//        }
//
//    }
//    println(count)
}

fun union(a: Int, b: Int) {
    val a = find(a)
    val b = find(b)

    if (know_truth[a] && know_truth[b]) return

    if (know_truth[a]) parent[b] = a
    else if (know_truth[b]) parent[a] = b

    else {
        if (a < b) parent[b] = a
        else parent[a] = b
    }
}

fun find(x: Int): Int {
    if (x != parent[x]) {
        parent[x] = find(parent[x])
    }

    return parent[x]
}
